src/Archives/BitStream.cpp

```diff
@@ -1,31 +1,29 @@
 #include "BitStream.h"
+#include <stdexcept>
+#include <string>
 
 namespace Archives
 {
-	BitStream::BitStream()
+	BitStream::BitStream() : 
+		m_BufferBitSize(0),
+		m_BufferSize(0),
+		m_Buffer(0), 
+		m_ReadBitIndex(0), 
+		m_WriteBitIndex(0),
+		m_WriteBuff(0) { }
+
+	BitStream::BitStream(std::size_t bufferSize, void *buffer) : 
+		m_BufferBitSize(bufferSize << 3),
+		m_BufferSize(bufferSize),
+		m_Buffer((unsigned char*)buffer),
+		m_ReadBitIndex(0),
+		m_WriteBitIndex(0),
+		m_WriteBuff(0)
 	{
-		// Initialize class variables
-		m_BufferSize = 0;
-		m_BufferBitSize = 0;
-		m_Buffer = 0;
-
-		m_ReadBitIndex = 0;
-		m_WriteBitIndex = 0;
-
-		m_WriteBuff = 0;
-	}
-
-	BitStream::BitStream(std::size_t bufferSize, void *buffer)
-	{
-		// Initialize class variables
-		m_BufferSize = bufferSize;
-		m_BufferBitSize = bufferSize << 3;
-		m_Buffer = (unsigned char*)buffer;
-
-		m_ReadBitIndex = 0;
-		m_WriteBitIndex = 0;
-
-		m_WriteBuff = 0;
+		// Check if storing bufferSize in bits wraps past largest value of a std::size_t
+		if (m_BufferBitSize < m_BufferSize) {
```

Comment:
This check will probably work, but it seems a bit obscure. I think I would prefer:
```cpp
if (bufferSize > std::numeric_limits<typename m_BufferBitSize>::max() / 8)
```

Reply:
Saying my code appears obscure seems like an odd role reversal. I liked it because it didn't require any division and was shorter :).

Including the limits header with the block of code shown above produces the following error on MSVC:

Error (active) E0260 explicit type is missing ('int' assumed) OP2Utility c:\ . . . \OP2Archive\OP2Utility\src\Archives\BitStream.cpp 26

and

Error (active) E0439 expected a '>'	OP2Utility c:\ . . . \OP2Archive\OP2Utility\src\Archives\BitStream.cpp 26

Reply:
Hmm, seems I gave you bad code. Guess I should have tested the idea first. The correction is:
```cpp
if (bufferSize > std::numeric_limits<decltype(m_BufferBitSize)>::max() / 8)
```

----

Indeed yes, your code is shorter, and uses very simple language features. My concern is that shorter isn't always easier to understand. In particular, it is an indirect test. It appears to be testing something which shouldn't possibly be true. It works because of overflow, which is not a super obvious thing to rely on. Worth mentioning, overflow is only well defined for `unsigned` values, so you need to know the type to ensure it is correct.

Using `numeric_limits`, the test doesn't rely on overflow to work. It's a more direct test as to the max buffer size that can be supported.